### PR TITLE
ENH: Halos from COLOSSUS joint mass-redshift distributions

### PR DESCRIPTION
**Pull request by rrjbca**
## Description
Implements `colossus_mf_redshift` and `colossus_mf` allowing users to (jointly) sample the redshifts (and masses) of dark matter halos for any of the mass functions in COLOSSUS. The structure of this new code mirrors the Schechter luminosity/mass functions in the galaxies module.

## Checklist
- [x] Follow the [Contributor Guidelines](https://github.com/skypyproject/skypy/blob/main/CONTRIBUTING.rst)
- [x] Write unit tests
- [x] Write documentation strings
- [x] Assign someone from your working team to review this pull request
- [x] Assign someone from the infrastructure team to review this pull request

### CODE REVIEW

File: skypy/halos/tests/test_colossus.py
Comment (Lucia-Fonseca) [inline skypy/halos/tests/test_colossus.py:57]: I get this warning, using Planck15 instead. Is this a problem?
`WARNING: Astropy cosmology class contains massive neutrinos, which are not taken into account in Colossus.`

File: skypy/halos/_colossus.py
Comment (Lucia-Fonseca) [inline skypy/halos/_colossus.py:135]: I am confused about the h units.
The colossus [massFunction](https://bdiemer.bitbucket.io/colossus/lss_mass_function.html#lss.mass_function.massFunction)  takes the mass parameters in units of `M_sun/h`. I understand `skypy.galaxies.redshift.redshifs_from_comoving_density` takes density in units of Mpc^-3.

My first idea was to do `lnmmin = np.log(m_min)` & `lnmmax = np.log(m_max)`, instead. But if I do that, the number of halos (after integrating for the given area of sky) does not agree with `len(z-halo)`

File: skypy/halos/_colossus.py
Comment (Lucia-Fonseca) [inline skypy/halos/_colossus.py:154]:
```suggestion
        is evaluated. Must be a monotonically increasing one-dimensional array
```

File: skypy/halos/_colossus.py
Comment (Lucia-Fonseca) [inline skypy/halos/_colossus.py:164]:
```suggestion
        Sky area over which halos are sampled. Must be in units of solid angle, steradian.
```

File: skypy/halos/tests/test_colossus.py
Comment (Lucia-Fonseca) [inline skypy/halos/tests/test_colossus.py:101]: Same Warning here `WARNING: Astropy cosmology class contains massive neutrinos, which are not taken into account in Colossus.`

File: skypy/halos/_colossus.py
Comment (Lucia-Fonseca) [inline skypy/halos/_colossus.py:195]: Why are we doing this? Instead of
```suggestion
    # Calculate the redshift at the centre of each bin
    z_mid = 0.5 * (redshift[:-1] + redshift[1:])
```

File: skypy/halos/redshift.py
Comment (Lucia-Fonseca) [inline skypy/halos/redshift.py:19]: I don't understand the purpose of this file.

**Review by Lucia-Fonseca — CHANGES_REQUESTED** [review]
* Some typos in doc-strings, 
* warnings, issue with Planck18
* and some questions

File: skypy/halos/_colossus.py
Comment (rrjbca) [reply skypy/halos/_colossus.py:15]: The functions are added to the `skypy.halos` and `skypy.halos.redshift` submodules instead

File: skypy/halos/_colossus.py
Comment (rrjbca) [reply skypy/halos/_colossus.py:164]: The units don't have to be steradians, any valid solid angle unit will be handled consistently

File: skypy/halos/tests/test_colossus.py
Comment (rrjbca) [reply skypy/halos/tests/test_colossus.py:57]: This is a colossus warning. Their power spectrum implementation doesn't account for massive neutrinos. It's not something we have any control over.

File: skypy/halos/tests/test_colossus.py
Comment (rrjbca) [reply skypy/halos/tests/test_colossus.py:42]: I'll use Planck15 instead

File: skypy/halos/redshift.py
Comment (rrjbca) [reply skypy/halos/redshift.py:19]: This mirrors the module structure used for the galaxy luminosity and stellar mass functions

File: skypy/halos/_colossus.py
Comment (rrjbca) [reply skypy/halos/_colossus.py:195]: It's just a different "representative" redshift for the bin based on the midpoint in comoving distance rather than redshift. Perhaps a better option would be the mean redshift of galaxies within the bin?

**Review by ntessore — CHANGES_REQUESTED** [review]
- As we discussed with the galaxies module, the splitting into `redshift` and `<quantity>` functions is a relic of when we could not multiassign columns. There should only be a single function that returns redshifts and properties.
- The output shape is `(2, nhalos)`. As we discussed around the multiassignment, we want functions to return `(nrows, ncolumns)`.

File: skypy/halos/_colossus.py
Comment (rrjbca) [reply skypy/halos/_colossus.py]: All of our function arguments and return value are in units of Msun, while the arguments for colossus massFunction need to be in units of Msun. I realise I messed up the conversion and this should be fixed now.

**Comment by rrjbca** [comment]
> * As we discussed with the galaxies module, the splitting into `redshift` and `<quantity>` functions is a relic of when we could not multiassign columns. There should only be a single function that returns redshifts and properties.

In that case (where) should `colossus_mf_redshift` appear in the documentation?

> * The output shape is `(2, nhalos)`. As we discussed around the multiassignment, we want functions to return `(nrows, ncolumns)`.

I'm not sure yet if that is the best way to address multiassignment; for anybody not using the Pipeline class it is natural for a function to return a tuple of different properties. Regardless I would rather merge as is than have this PR held up waiting for multiassignment to be updated on the main branch.

**Comment by ntessore** [comment]
> In that case (where) should `colossus_mf_redshift` appear in the documentation?

There should only be one function. It goes through everything sequentially anyway.

> I'm not sure yet if that is the best way to address multiassignment; for anybody not using the Pipeline class it is natural for a function to return a tuple of different properties. Regardless I would rather merge as is than have this PR held up waiting for multiassignment to be updated on the main branch.

I don't want to have another API in the codebase that we already know needs changing. And I think you once pointed out to me that multiassignment assumes `(nrows, ncolumns)` and it's by hack/coincidence that the other way round works, too.

I was very much in the "columns like tuples" camp, but I am now very much in the "columns as columns" camp, because you cannot easily stack the former.

File: skypy/halos/_colossus.py
Comment (Lucia-Fonseca) [reply skypy/halos/_colossus.py]: I am not sure of that. I can't make `n_halo` and `len(z_halo)` agree with your new changes.

Comment (rrjbca) [reply skypy/halos/_colossus.py]: There is a unit test checking exactly that and passing though? `m_min` and `m_max` are in units of M_sun and `massFunction` takes masses in units of M_sun/h so you have to do a change of units somewhere; I chose to do the integral in units of M_sun/h and therefore rescaled the limits by h.

**Comment by rrjbca** [comment]
I am strongly against working in this way in general. Pull requests can only realistically be reviewed in the context of the code on the base branch (and/or main branch). What you're suggesting leaves me stuck waiting for the multiassignment "fix" before this PR can be merged and I can carry on with downstream work. I imagine this would be particularly frustrating for an external contributor being made to wait on a separate issue through no fault of their own.

**Comment by rrjbca** [comment]
Sorry, I thought (ii) in you suggestion also required a change to `Pipeline` hence my pushback.

File: skypy/halos/_colossus.py
Comment (Lucia-Fonseca) [reply skypy/halos/_colossus.py]: Strange. That happens when I run the unit test locally on my computer.

File: skypy/halos/_colossus.py
Comment (Lucia-Fonseca) [reply skypy/halos/_colossus.py:195]: Perhaps. It's not clear to me why to choose this representative redshift over something else, e.g. what you suggest about the mean redshift within the bin or my suggestion.

**Review by ntessore — APPROVED** [review]
After some local discussion, we agreed to merge this as is and then try to sort out the questions across the codebase separately.

**Review by Lucia-Fonseca — APPROVED** [review]

**Pull request merged**
